src/lib.rs

```diff
@@ -128,65 +196,95 @@ impl DstService {
-
+                    if curr.is_empty() {
                         tx.send(Ok(pb::Update {
-                            update: Some(pb::update::Update::Add(pb::WeightedAddrSet {
-                                addrs: added,
-                                metric_labels,
+                            update: Some(pb::update::Update::NoEndpoints(pb::NoEndpoints {
+                                exists: true,
                             })),
                         }))
-                        .await?;
-                    }
+                        .await?
+                    } else {
+                        let added = curr
+                            .iter()
+                            .filter(|(_, ep)| Self::is_add(&prev, ep))
+                            .map(|(addr, meta)| {
+                                let protocol_hint = if meta.h2 {
+                                    Some(pb::ProtocolHint {
+                                        protocol: Some(pb::protocol_hint::Protocol::H2(
+                                            pb::protocol_hint::H2::default(),
+                                        )),
+                                    })
+                                } else {
+                                    None
+                                };
+
+                                let mut metric_labels = HashMap::default();
+                                metric_labels.insert("addr".to_string(), addr.to_string());
+                                metric_labels.insert("h2".to_string(), meta.h2.to_string());
+                                metric_labels.extend(meta.metric_labels.clone());
+
+                                pb::WeightedAddr {
+                                    addr: Some(addr.into()),
+                                    weight: meta.weight,
+                                    protocol_hint,
+                                    metric_labels,
+                                    tls_identity: meta.tls_identity.clone().map(|name| {
+                                        pb::TlsIdentity {
+                                            strategy: Some(
+                                                pb::tls_identity::Strategy::DnsLikeIdentity(
+                                                    pb::tls_identity::DnsLikeIdentity { name },
+                                                ),
+                                            ),
+                                        }
+                                    }),
+                                    authority_override: meta.authority_override.clone().map(
+                                        |authority_override| pb::AuthorityOverride {
+                                            authority_override,
+                                        },
+                                    ),
+                                }
+                            })
```

Comment:
take it or leave it: could be kind of nice to factor out the conversion of `(meta, addr)`s into `pb::WeightedAddr`s into a separate function? this is kind of a big pile of code.